assert.h

```diff
@@ -21,19 +24,31 @@
 
 namespace snowhouse
 {
+  auto WithMessage = [](const std::string& msg) { return [&]() { return msg; }; };
```

Comment:
I was thinking `With*` can be awkward because matchers use it:

```cpp
Assert::That("a", Is().StartingWith("b"), WithMessage("Invalid result for input: a"));
```

I agree with you that `MessageOnFailure` can be confusing since it sounds like it replaces the entire failure message.

Some other ideas

- `FailureContext`, `FailureReason`
- `ExtraMessage`
- `Explanation`

Reply:
+1 for `ExtraMessage`

Reply:
Cool, let's use `ExtraMessage` then.